src/volume_mgr.rs

```diff
@@ -866,6 +864,146 @@ where
+                    // update the parent dir with the cluster of the new dir
+                    fat.write_entry_to_disk(&self.block_device, &new_dir_entry_in_parent)?;
+                }
+                let new_dir_start_block = fat.cluster_to_block(new_dir_entry_in_parent.cluster);
+                debug!("Made new dir entry {:?}", new_dir_entry_in_parent);
+                let now = self.time_source.get_timestamp();
+                let fat_type = fat.get_fat_type();
+                // A blank block
+                let mut blocks = [Block::new()];
+                // make the "." entry
+                let dot_entry_in_child = DirEntry {
+                    name: crate::ShortFileName::this_dir(),
+                    mtime: now,
+                    ctime: now,
+                    attributes: att,
+                    // point at ourselves
+                    cluster: new_dir_entry_in_parent.cluster,
+                    size: 0,
+                    entry_block: new_dir_start_block,
+                    entry_offset: 0,
+                };
+                debug!("New dir has {:?}", dot_entry_in_child);
+                let mut offset = 0;
+                blocks[0][offset..offset + OnDiskDirEntry::LEN]
+                    .copy_from_slice(&dot_entry_in_child.serialize(fat_type)[..]);
+                offset += OnDiskDirEntry::LEN;
+                // make the ".." entry
+                let dot_dot_entry_in_child = DirEntry {
+                    name: crate::ShortFileName::parent_dir(),
+                    mtime: now,
+                    ctime: now,
+                    attributes: att,
+                    // point at our parent
+                    cluster: match fat_type {
+                        FatType::Fat16 => {
+                            // On FAT16, indicate parent is root using Cluster(0)
+                            if parent_directory_info.cluster == ClusterId::ROOT_DIR {
+                                ClusterId::EMPTY
+                            } else {
+                                parent_directory_info.cluster
+                            }
+                        }
+                        FatType::Fat32 => parent_directory_info.cluster,
+                    },
+                    size: 0,
+                    entry_block: new_dir_start_block,
+                    entry_offset: OnDiskDirEntry::LEN_U32,
+                };
+                debug!("New dir has {:?}", dot_dot_entry_in_child);
+                blocks[0][offset..offset + OnDiskDirEntry::LEN]
+                    .copy_from_slice(&dot_dot_entry_in_child.serialize(fat_type)[..]);
+
+                self.block_device
+                    .write(&blocks, new_dir_start_block)
+                    .map_err(Error::DeviceError)?;
+
+                // Now zero the rest of the cluster
+                for b in blocks[0].iter_mut() {
+                    *b = 0;
+                }
```

Comment:
Shouldn't this be the other way around?
```suggestion
                // Now zero the rest of the cluster
                for b in blocks[0].iter_mut() {
                    *b = 0;
                }
                
                self.block_device
                    .write(&blocks, new_dir_start_block)
                    .map_err(Error::DeviceError)?;
```

Reply:
You I that it wouldn't write the two directory entries I just carefully put into that block. Besides, it was zero initialised.

Having written the first block in the cluster, I need to write zeros to the rest of the cluster (which is up to 32 blocks long). I already had a block on the stack so I zeroed it. I could make a second block, but that probably just chews up more stack space.

Reply:
You are of course right. I was dumb and didn't read that code carefully, and somehow thought "Now zero the rest of the cluster" was about clearing the rest of `blocks[0]` before it gets written to the card. But that doesn't make sense. As you said, it was zero initialized from the beginning anyway.